File: src/shellext/osx_finder/src/AeroNode.m
#import "AeroNode.h"

@interface AeroNode (Private)
-(void)updateChildrenStatus;
-(AeroNode*)nodeAtPath:(NSArray*)pathComponents arrayIndex:(int)index createPath:(BOOL)create;
@end

@implementation AeroNode

@synthesize name;
@synthesize parent;
@synthesize children;
@synthesize ownStatus;
@synthesize childrenStatus;

-(void)dealloc
{
    [name release];
    [children release]; // will release all sub-nodes
    [super dealloc];
}

- (id)init
{
    // make it impossible to call init by calling initWithName with an empty name
    return [self initWithName:@"" andParent:nil];
}

-(id)initWithName:(NSString*)theName andParent:(AeroNode*)theParent
{
    self = [super init];
    if (self) {

        // Do not create Nodes with empty names
        if (theName == nil || theName.length == 0) {
            [self release];
            [super doesNotRecognizeSelector:_cmd];
            return nil;
        }

        children = [[NSMutableDictionary alloc] init];
        name = [theName retain];
        parent = theParent;
        ownStatus = NoStatus;
        childrenStatus = NoStatus;
    }

    return self;
}

-(NodeStatus)status
{
    return ownStatus | childrenStatus;
}

-(void)setStatus:(NodeStatus)theStatus
{
    ownStatus = theStatus;
    [parent updateChildrenStatus];
}

/**
 Update the node's childrenStatus field with the status of all children
 If the status has changed, recursively call the parent's updateChildrenStatus
 */
-(void) updateChildrenStatus
{
    NodeStatus oldStatus = self.status;

    childrenStatus = NoStatus;
    NSEnumerator *enumerator = [children objectEnumerator];
    NSMutableArray* childrenToDelete = [NSMutableArray arrayWithCapacity:[children count]];
    AeroNode* child;
    while ((child = [enumerator nextObject])) {
        NodeStatus st = child.status;
        childrenStatus |= st;
        if (st == NoStatus) {
            [childrenToDelete addObject:child.name];
        }
    }

    [children removeObjectsForKeys:childrenToDelete];

    if (self.status != oldStatus) {
        [parent updateChildrenStatus];
    }
}

/**
 Returns a pointer to a AeroNode at the given path.
 If create is set to YES, creates the intermediary nodes if they don't exist,
 (similarly to mkdir -p), otherwise, returns nil.
 */
-(AeroNode*) getNodeAtPath:(NSString*)path createPath:(BOOL) create
{
    NSString* root = self.name;

    if (![path hasPrefix:root]) {
        return nil; // path is not under root
    }

    if (path.length == root.length) {
        return self; // path and root are the same
    }

    if ([path characterAtIndex:root.length] != '/') {
        return nil; // ensure path is indeed under root (there's a path separator)
    }

    // Remove root from the path and get an array with the other path components
    path = [path substringFromIndex:root.length + 1]; // +1 to remove the path separator
    NSArray* pathComponents = [path pathComponents];

    if (pathComponents.count == 0) {
        return self;
    }

    return [self nodeAtPath:pathComponents arrayIndex:0 createPath:create];
}

-(AeroNode*) nodeAtPath:(NSArray*)pathComponents arrayIndex:(int)index createPath:(BOOL)create
{
    if (index == [pathComponents count]) {
        return self;
    }

    NSString* component = [pathComponents objectAtIndex:index];
    AeroNode* child = [children objectForKey:component];

    if (child == nil) {
        if (!create) {
            return nil;
        }
        child = [[AeroNode alloc] initWithName:component andParent:self];
        [children setObject:child forKey:component];
        [child release];
    }

    index++;
    return [child nodeAtPath:pathComponents arrayIndex:index createPath:create];
}

@end

File: src/shellext/win_explorer/src/ShellExtension/AeroNode.cpp
#include "stdafx.h"
#include "AeroNode.h"
#include <assert.h>
#include <algorithm>
#include <list>

#include <Shlobj.h>

#include "logger.h"
#include "string_helpers.h"

AeroNode::AeroNode(const std::wstring& name, AeroNode* parent)
	: m_name(lowercase(name)),
	m_parent(parent),
	m_children(),
	m_ownStatus(NoStatus),
	m_childrenStatus(NoStatus)
{
	assert(!m_name.empty());
}

AeroNode::~AeroNode()
{
	// Delete all children nodes
	for (auto it = m_children.begin(); it != m_children.end(); ++it) {
		delete it->second;
	}
}


/**
 * Returns a pointer to a AeroNode at the given path.
 * If createPath is true, creates the intermediary nodes if they don't exist (like 'mkdir -p')
 * Note: if createPath is true, this method could still return NULL if the path doesn't match the root node
*/
AeroNode* AeroNode::nodeAtPath(const std::wstring& p, bool createPath)
{
	std::wstring root = m_name;;
	std::wstring path = lowercase(p);

	if (!str_starts_with(path, root)) {
		return NULL; // path is not under root
	}

	if (path.length() == root.length()) {
		return this; // path and root are the same
	}

	if (path.at(root.length()) != L'\\') {
		return NULL; // ensure path is indeed under root (there's a path separator)
	}

	// Remove root from the path and get a vector with the other path components
	path = path.substr(root.length() + 1); // +1 to remove the path separator
	std::vector<std::wstring> pathComponents = split(path, L'\\');

	if (pathComponents.empty()) {
		// edge case where path equals root plus one or more trailing path separators
		return this;
	}

	return nodeAtPath(pathComponents, 0, createPath);
}

/**
 * Internal helper function that is called recursively to find the node for a given path
 * @param: pathComponents: a vector with path components
 * index: current position in the path component vector. We increase this variable as we recurse.
 * if true, create intermediate nodes in the path if they don't exist
 */
AeroNode* AeroNode::nodeAtPath(const std::vector<std::wstring>& pathComponents, int index, bool create)
{
	if (index == pathComponents.size()) {
		assert(m_name == pathComponents[index - 1]);
		return this;
	}

	const std::wstring& component = pathComponents[index];
	AeroNode* child;
	auto it = m_children.find(component);
	if (it == m_children.end()) {
		if (!create) {
			return NULL;
		}
		child = new AeroNode(component, this);
		m_children[component] = child;
	} else {
		child = it->second;
	}

	index++;
	return child->nodeAtPath(pathComponents, index, create);
}

AeroNode::Status AeroNode::status() const
{
	return m_ownStatus | m_childrenStatus;
}

void AeroNode::setStatus(Status status)
{
	m_ownStatus = status;

	std::wstring path = pathOfNode();
	DEBUG_LOG("Notifying the shell of overlay change: " << path);
	SHChangeNotify(SHCNE_UPDATEITEM, SHCNF_PATH | SHCNF_FLUSHNOWAIT, path.c_str(), NULL);

	if (m_parent) {
		m_parent->updateChildrenStatus();
	}
}

/**
Update the node's childrenStatus field with the status of all children
If the status has changed, recursively call the parent's updateChildrenStatus
*/
void AeroNode::updateChildrenStatus()
{
	Status oldStatus = status();

	m_childrenStatus = NoStatus;

	auto iter = m_children.begin();
	while (iter != m_children.end()) {
		AeroNode* child = iter->second;
		Status st = child->status();
		m_childrenStatus |= st;
		if (st == NoStatus) {
			m_children.erase(iter++);
			delete child;
		} else {
			++iter;
		}
	}

	if (status() != oldStatus) {
		std::wstring path = pathOfNode();
		DEBUG_LOG("Notifying the shell of overlay change: " << path);
		SHChangeNotify(SHCNE_UPDATEITEM, SHCNF_PATH | SHCNF_FLUSHNOWAIT, path.c_str(), NULL);

		if (m_parent) {
			m_parent->updateChildrenStatus();
		}
	}
}

std::wstring AeroNode::name() const
{
	return m_name;
}

AeroNode* AeroNode::parent() const
{
	return m_parent;
}

std::map<std::wstring, AeroNode*> AeroNode::children()
{
	return m_children;
}

/**
 * Return the full path of this node
 */
std::wstring AeroNode::pathOfNode() const
{
	// Create a list of all parents
	const AeroNode*  node = this;
	std::list<const AeroNode*> parents;
	do {
		parents.push_front(node);
	} while (node = node->parent());

	// Write the name of all parents in a string
	std::wstring path;
	path.reserve(MAX_PATH);
	for(auto iter = parents.begin(); iter != parents.end(); ++iter) {
		path += (*iter)->name() + L'\\';
	}

	// drop the trailing slash
	path.resize(path.length() - 1);

	return path;
}
